base/asyncmap.jl

```diff
@@ -145,4 +145,27 @@ function next(itr::StreamMapIterator, state::StreamMapState)
     return (r, state)
 end
 
-iteratorsize(::Type{StreamMapIterator}) = SizeUnknown()
+iteratorsize(::Type{StreamGenerator}) = SizeUnknown()
+
+
+"""
+    asyncgenerate(f, c...) -> iterator
+
+Apply `@async f` to each element of `c`.
+
+For multiple collection arguments, apply f elementwise.
+
+Results are returned in order as they become available.
+"""
+asyncgenerate(f, c...) = StreamGenerator(f, c...)
```

Comment:
I started along those lines, but ended up concluding that the implementation just becomes needlessly complicated if the two layers are merged. i.e. it makes sense to have two distinct architectural layers in the implementation. One runs stuff asynchronously and collects the results in whatever order they happen to be computed, the other manages in-order iteration over the results collection. I am open to suggestions on naming. However, these names are not exported, and the basic design was reviewed and accepted already in #15058. I suggest that these names are good enough.

[This line](https://github.com/JuliaLang/julia/pull/15409/commits/c6bc3680d15d9b5762d6c7a1fad62d765247285e#diff-9d7ce9b013c358a3f6e8126bbdc03145R179) demonstrates how the `AsyncCollector` is useful on its own.

I agree that the API could be made simpler, I've commented on that over here: https://github.com/JuliaLang/julia/issues/14843#issuecomment-202656425 
But, I think in this instance the implementation is simpler with `AsyncCollector` kept seperate from `AsyncGenerator`.